File: Tools/TanhSaturation.m
function y = TanhSaturation(x, ksigma)
% Saturates outlier samples using a tanh shape function
% For matrix data in the form of channels x time, the data is saturated over the second
% dimention (time)
%
% Open Source ECG Toolbox, version 3.14, Jan 2020
% Reza Sameni
% Jan 2020
%
% Revised:
%   Nov 2020: simplified formula

alpha = ksigma * std(x, [], 2);
alpha = alpha(:);
y = diag(alpha) * tanh(diag(1./alpha) * x);

% sd = std(x, [], 2);
% alpha = ksigma * sd;
% T = size(x, 2);
% y = alpha(:, ones(1, T)).*tanh(x./alpha(:, ones(1, T)));

File: python/src/oset/generic/tanh_saturation.py
import argparse
import numpy as np


def tanh_saturation(x, ksigma):
    """
    Saturates outlier samples using a tanh shape function.

    Args:
        x (ndarray): Input data, can be a vector or a matrix (channels x time).
        ksigma (float): Scaling factor for the saturation level.

    Returns:
        ndarray: Saturated data with outliers replaced by the saturation level.

    Revision History:
        2023: Translated to Python from Matlab

    Amulya Jain, 2023
    The Open-Source Electrophysiological Toolbox
    https://github.com/alphanumericslab/OSET
    """
    # Compute the scaling factor based on the standard deviation of each channel
    alpha = ksigma * np.std(x, axis=0, keepdims=True)

    # Convert to a column vector
    alpha = alpha.flatten()

    # Scale the input data and apply the tanh function to saturate outliers
    y = (np.diag(alpha) * np.tanh(np.diag(1.0 / alpha) * x))[0]
    return y


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="""
    Saturates outlier samples using a tanh shape function.

    Args:
        x (ndarray): Input data, can be a vector or a matrix (channels x time).
        ksigma (float): Scaling factor for the saturation level.

    Returns:
        ndarray: Saturated data with outliers replaced by the saturation level.

    Revision History:
        2023: Translated to Python from Matlab

    """,
        formatter_class=argparse.RawTextHelpFormatter,
    )
    args = parser.parse_args()
